### PR TITLE
Level DB Benchmark Tool and Call Only Mode

### PR DESCRIPTION
**Pull request by praetoriansentry**
# Description
Adding a command to assist in the load testing of LevelDB. Also adding a mode to polycli to add a mode for specifically running `eth_call` variations rather than actual transactions which helps in the load testing of rpc nodes on mainnets.

image

image

# Testing

- [ ] Concurrent run of load test in random mode
- [ ] Concurrent run of load test in random mode + call only
- [ ] Concurrent run of load test in random mode + call only + pre-deploy all contracts
- [ ] Concurrent run of Recall mode in empty chain
- [ ] Concurrent run of Recall mode in populated chain

### CODE REVIEW

File: cmd/leveldbbench/leveldbbench.go [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+func runFullCompact(ctx context.Context, db *leveldb.DB, wo *opt.WriteOptions) {
+	err := db.CompactRange(util.Range{Start: nil, Limit: nil})
+	if err != nil {
+		log.Fatal().Err(err).Msg("error compacting data")
+	}
+}
+func runFullScan(ctx context.Context, db *leveldb.DB, wo *opt.WriteOptions, ro *opt.ReadOptions) (uint64, []uint64) {
+	pool := make(chan bool, *degreeOfParallelism)
+	var wg sync.WaitGroup
+	// 32 should be safe here. That would correspond to a single value that's 4.2 GB
```
Comment (IdrisHanafi) [inline cmd/leveldbbench/leveldbbench.go:261]: > That would correspond to a single value that's 4.2 GB

curious on what this comment means 👀

File: cmd/leveldbbench/leveldbbench.go [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+
+	// if that key doesn't exist exactly advance to the next key
+	if !exists {
+		exists = r.iterator.Next()
+	}
+	// if there is no next key, to back to the beginning
+	if !exists {
+		r.iterator.First()
+		r.iterator.Next()
+	}
```
Comment (IdrisHanafi) [inline cmd/leveldbbench/leveldbbench.go:434]: curious about how good of a random this is. how often would the initial random key be at a point where there is no keys next in line and having this method return the first key:
```
// first try to just get a random key
exists := r.iterator.Seek(seekKey)
```

File: cmd/leveldbbench/leveldbbench.go [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+func init() {
+	flagSet := LevelDBBenchCmd.PersistentFlags()
+	writeLimit = flagSet.Uint64("write-limit", 1000000, "The number of entries to write in the db")
+	readLimit = flagSet.Uint64("read-limit", 10000000, "the number of reads will attempt to complete in a given test")
+	overwriteCount = flagSet.Uint64("overwrite-count", 5, "the number of times to overwrite the data")
+	sequentialReads = flagSet.Bool("sequential-reads", false, "if true we'll perform reads sequentially")
+	sequentialWrites = flagSet.Bool("sequential-writes", false, "if true we'll perform writes in somewhat sequential manner")
+	keySize = flagSet.Uint64("key-size", 32, "The byte length of the keys that we'll use")
+	degreeOfParallelism = flagSet.Uint8("degree-of-parallelism", 2, "The number of concurrent goroutines we'll use")
+	rawSizeDistribution = flagSet.String("size-distribution", borDistribution, "the size distribution to use while testing")
```
Comment (IdrisHanafi) [inline cmd/leveldbbench/leveldbbench.go:606]: nice, so i can technically generate the distribution for any client and use it here

File: cmd/loadtest/loadtest.go
Comment (IdrisHanafi) [inline cmd/loadtest/loadtest.go:1]: bless ur soul for refactoring this 🙏

File: cmd/loadtest/loadtest.go [hunk 4]
```diff
@@ -1100,12 +869,23 @@ func loadtestCall(ctx context.Context, c *ethclient.Client, nonce uint64, ltCont
...
+		msg := txToCallMsg(tx)
+		_, err = c.CallContract(ctx, msg, nil)
+	} else {
+		_, err = contracts.CallLoadTestFunctionByOpCode(f, ltContract, tops, *iterations)
+	}
 	return
 }
 
-func loadtestCallPrecompiledContracts(ctx context.Context, c *ethclient.Client, nonce uint64, ltContract *contracts.LoadTester, useSelectedAddress bool) (t1 time.Time, t2 time.Time, err error) {
+func loadTestCallPrecompiledContracts(ctx context.Context, c *ethclient.Client, nonce uint64, ltContract *contracts.LoadTester, useSelectedAddress bool) (t1 time.Time, t2 time.Time, err error) {
```
Comment (IdrisHanafi) [inline cmd/loadtest/loadtest.go:888]: seeing this again reminds me of the discussion of boolean flags haha

File: cmd/loadtest/recall.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+func getRecentBlocks(ctx context.Context, ec *ethclient.Client, c *ethrpc.Client) ([]*json.RawMessage, error) {
+	bn, err := ec.BlockNumber(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rawBlocks, err := util.GetBlockRange(ctx, bn-*inputLoadTestParams.RecallLength, bn, c)
+	return rawBlocks, err
+}
+
+func getRecallTransactions(ctx context.Context, c *ethclient.Client, rpc *ethrpc.Client) ([]rpctypes.PolyTransaction, error) {
```
Comment (IdrisHanafi) [inline cmd/loadtest/recall.go:24]: so this fxn just returns all of the transactions of a given block range

File: cmd/leveldbbench/usage.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+     Kbytes              : count     distribution
+         0 -> 1          : 0        |                                        |
+         2 -> 3          : 0        |                                        |
+         4 -> 7          : 10239    |****************                        |
+         8 -> 15         : 25370    |****************************************|
+        16 -> 31         : 7082     |***********                             |
+        32 -> 63         : 1241     |*                                       |
+        64 -> 127        : 58       |                                        |
+       128 -> 255        : 11       |                                        |
+```
```
Comment (IdrisHanafi) [inline cmd/leveldbbench/usage.md:22]: nice so this is the baseline for a synced bor node

**Review by IdrisHanafi — APPROVED** [review]
lgtm! just left minor questions

File: cmd/leveldbbench/leveldbbench.go (hunk 1 above)
Comment (praetoriansentry) [reply cmd/leveldbbench/leveldbbench.go:261]: We are basically taking the length of the value then getting the bit length of that length. So a bit length of 32 supports an underlying value of 4gb which is larger than I suspect we need. But we could always increase.

File: cmd/leveldbbench/leveldbbench.go (hunk 2 above)
Comment (praetoriansentry) [reply cmd/leveldbbench/leveldbbench.go:434]: It largely depends on the number of keys in the underlying level db. If you have 100, then I suspect (I could be wrong) that 1/100 times you'll end up with not greater keys and you just go back to the beginning. 

Where this goes wrong is when the actual keys aren't random. Eg in geth - many of the keys are prefixed with an ascii string like "h". In that case the random won't be as good.

File: cmd/leveldbbench/leveldbbench.go (hunk 3 above)
Comment (praetoriansentry) [reply cmd/leveldbbench/leveldbbench.go:606]: I hope. The idea is to hard code what we found in bor but support other distributions easily.

File: cmd/loadtest/loadtest.go
Comment (praetoriansentry) [reply cmd/loadtest/loadtest.go:1]: It was time. And still more to do of course.

File: cmd/loadtest/loadtest.go (hunk 4 above)
Comment (praetoriansentry) [reply cmd/loadtest/loadtest.go:888]: Yeah. Sometimes it's ugly but the easiest way to get it done. I didn't look into this one specifically.

File: cmd/loadtest/recall.go (hunk 5 above)
Comment (praetoriansentry) [reply cmd/loadtest/recall.go:24]: Yeah ideally. The goal being to replay this transactions.

File: cmd/leveldbbench/usage.md (hunk 6 above)
Comment (praetoriansentry) [reply cmd/leveldbbench/usage.md:22]: Yeah. What's weird is this is quite different than the underlying level db key sizes. There are several layers of abstraction between leveldb api and underlying disk IOPs, so I should t be surprised.

File: cmd/leveldbbench/leveldbbench.go (hunk 2 above)
Comment (praetoriansentry) [reply cmd/leveldbbench/leveldbbench.go:434]: I think if we wanted to be more accurate about it, need to take some kind of sample of the actual keys. Maybe a future improvement.

**Pull request merged**